### PR TITLE
Implement Logging of Self-Feedback in logs/Debug Folder

### PR DESCRIPTION
**Pull request by AndresCdo**
### Background
To improve the performance of our application and understand how it functions in different scenarios, we decided to implement self-feedback logging. This feature will enable the application to log its own feedback regarding its performance and decisions, providing valuable insights for future optimization and development.

### Changes
Developed a self-feedback logging mechanism that captures application feedback and appends it to the log file within the logs/DEBUG folder.

### Documentation
Descriptions were added in new lines on tests.

### Test Plan
I did test it, and it seems to work fine.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Comment by AndresCdo** [comment]
#3842

**Comment by merwanehamadi** [comment]
@AndresCdo do it like that: 

```
def get_self_feedback(self, thoughts: dict, llm_model: str) -> str:
        """Generates a feedback response based on the provided thoughts dictionary.
        This method takes in a dictionary of thoughts containing keys such as 'reasoning',
        'plan', 'thoughts', and 'criticism'. It combines these elements into a single
        feedback message and uses the create_chat_completion() function to generate a
        response based on the input message.
        Args:
            thoughts (dict): A dictionary containing thought elements like reasoning,
            plan, thoughts, and criticism.
        Returns:
            str: A feedback response generated using the provided thoughts dictionary.
        """
        ai_role = self.config.ai_role

        feedback_prompt = f"Below is a message from me, an AI Agent, assuming the role of {ai_role}. whilst keeping knowledge of my slight limitations as an AI Agent Please evaluate my thought process, reasoning, and plan, and provide a concise paragraph outlining potential improvements. Consider adding or removing ideas that do not align with my role and explaining why, prioritizing thoughts based on their significance, or simply refining my overall thought process."
        reasoning = thoughts.get("reasoning", "")
        plan = thoughts.get("plan", "")
        thought = thoughts.get("thoughts", "")
        feedback_thoughts = thought + reasoning + plan
        
        messages = {"role": "user", "content": feedback_prompt + feedback_thoughts}
        #log messages
        feedback = create_chat_completion(
            [{"role": "user", "content": feedback_prompt + feedback_thoughts}],
            llm_model,
        )
        #log feedback
        return  feedback
```

**Comment by AndresCdo** [comment]
> 1st log call it prompt_self_feedback.json
> 
> and the second log call it self_feedback.txt
image

**Comment by merwanehamadi** [comment]
@AndresCdo my bad I talked to the team I think prompt_supervisor.json and supervisor_feedback are better names, more agnostic. Self feedback assumes the agent criticizes itself.

**Comment by AndresCdo** [comment]
> @AndresCdo do it like that:
> 
> ```
> def get_self_feedback(self, thoughts: dict, llm_model: str) -> str:
>         """Generates a feedback response based on the provided thoughts dictionary.
>         This method takes in a dictionary of thoughts containing keys such as 'reasoning',
>         'plan', 'thoughts', and 'criticism'. It combines these elements into a single
>         feedback message and uses the create_chat_completion() function to generate a
>         response based on the input message.
>         Args:
>             thoughts (dict): A dictionary containing thought elements like reasoning,
>             plan, thoughts, and criticism.
>         Returns:
>             str: A feedback response generated using the provided thoughts dictionary.
>         """
>         ai_role = self.config.ai_role
> 
>         feedback_prompt = f"Below is a message from me, an AI Agent, assuming the role of {ai_role}. whilst keeping knowledge of my slight limitations as an AI Agent Please evaluate my thought process, reasoning, and plan, and provide a concise paragraph outlining potential improvements. Consider adding or removing ideas that do not align with my role and explaining why, prioritizing thoughts based on their significance, or simply refining my overall thought process."
>         reasoning = thoughts.get("reasoning", "")
>         plan = thoughts.get("plan", "")
>         thought = thoughts.get("thoughts", "")
>         feedback_thoughts = thought + reasoning + plan
>         
>         messages = {"role": "user", "content": feedback_prompt + feedback_thoughts}
>         #log messages
>         feedback = create_chat_completion(
>             [{"role": "user", "content": feedback_prompt + feedback_thoughts}],
>             llm_model,
>         )
>         #log feedback
>         return  feedback
> ```
> 
> in autogpt/agent/agent.py
> 
> 1st log call it prompt_self_feedback.json
> 
> and the second log call it self_feedback.txt

This logic breaks tests:
image

**Comment by AndresCdo** [comment]
image

File: autogpt/agent/agent.py [hunk 1]
```diff
@@ -327,7 +343,12 @@ def get_self_feedback(self, thoughts: dict, llm_model: str) -> str:
...
         thought = thoughts.get("thoughts", "")
         feedback_thoughts = thought + reasoning + plan
-        return create_chat_completion(
+
+        messages = {"role": "user", "content": feedback_prompt + feedback_thoughts}
+
+        feedback = create_chat_completion(
             [{"role": "user", "content": feedback_prompt + feedback_thoughts}],
             llm_model,
         )
```
Comment (waynehamadi) [inline autogpt/agent/agent.py:354]: oh you don't need to change the signature, 
1- just keep returning the feedback.
2- log the messages WITHIN the get_self_feedback method as well
3- log the feedback WITHIN the get_self_feedback method as well

And make sure your FIRST log the messages (prompt_supervisor_feedback.json) before you log the feedback (supervisor_feedback.txt)

**Review by merwanehamadi — CHANGES_REQUESTED** [review]

File: autogpt/agent/agent.py (hunk 1 above)
Comment (AndresCdo) [reply autogpt/agent/agent.py:354]: Yes, I did it before and tests were broken. Check it know, I did reset it.

File: autogpt/log_cycle/log_cycle.py [hunk 2]
```diff
@@ -10,6 +10,8 @@
 NEXT_ACTION_FILE_NAME = "next_action.json"
 PROMPT_SUMMARY_FILE_NAME = "prompt_summary.json"
 SUMMARY_FILE_NAME = "summary.txt"
+SELF_FEEDBACK_FILE_NAME = "supervisor_feedback.txt"
```
Comment (waynehamadi) [inline autogpt/log_cycle/log_cycle.py]: great !

@AndresCdo please call it SUPERVISOR_FEEDBACK_FILE_NAME

File: autogpt/log_cycle/log_cycle.py [hunk 3]
```diff
@@ -10,6 +10,8 @@
...
+PROMPT_FEEDBACK_FILE_NAME = "prompt_supervisor.json"
```
Comment (waynehamadi) [inline autogpt/log_cycle/log_cycle.py]: please call this one PROMPT_SUPERVISOR_FEEDBACK_FILE_NAME and prompt_supervisor_feedback.json

File: autogpt/agent/agent.py [hunk 4]
```diff
@@ -327,7 +329,27 @@ def get_self_feedback(self, thoughts: dict, llm_model: str) -> str:
...
+        self.log_cycle_handler.log_cycle(
+            self.config.ai_name,
+            self.created_at,
+            self.cycle_count,
+            messages,
+            PROMPT_FEEDBACK_FILE_NAME,
+        )
+
+        feedback = create_chat_completion(
             [{"role": "user", "content": feedback_prompt + feedback_thoughts}],
```
Comment (waynehamadi) [inline autogpt/agent/agent.py]: pass the messages variable here please, not :

[{"role": "user", "content": feedback_prompt + feedback_thoughts}],

**Comment by anonhostpi** [comment]
Some of the work here may serve as a good basis for #3916 and #15

**Review by merwanehamadi — CHANGES_REQUESTED** [review]

File: autogpt/agent/agent.py [hunk 5]
```diff
@@ -11,6 +11,8 @@
 from autogpt.log_cycle.log_cycle import (
     FULL_MESSAGE_HISTORY_FILE_NAME,
     NEXT_ACTION_FILE_NAME,
+    PROMPT_SUPERVISOR_FEEDBACK_FILE_NAME,
+    SUPERVISOR_FEEDBACK_FILE_NAME,
```
Comment (waynehamadi) [inline autogpt/agent/agent.py:17]: great ! now please fix the test that fails.
image

**Review by merwanehamadi — CHANGES_REQUESTED** [review]

**Comment by Boostrix** [comment]
> Some of the work here may serve as a good basis for https://github.com/Significant-Gravitas/Auto-GPT/issues/3916 and https://github.com/Significant-Gravitas/Auto-GPT/issues/15

Indeed, improving self feedback (#3934) could be used as "reward" inside a fitness function. Especially in combination with user input (human feedback):

- Any time the human isn't saying 'y' but something else, that should impact the fitness function. 
- And when inside continuous mode (#150), any "error/warning" should affect the reward, too. 
- Observing the number of preauthorised commands and their performance, should also affect the fitness function (#1035). 
- Human intervention by pausing (#3083), should negatively affect the reward. 
- Letting user feedback time out (#3971) should have NO impact on reward (the agent is probably running unattended or doing just fine). 
- The agent prompting for feedback (#1548 / #3396) but being told to just continue, should reinforce the current path. 
- Likewise increasing the number of preauthorized steps, should positively affect the reward: #2481

**Review by merwanehamadi — APPROVED** [review]

**Pull request merged**